### PR TITLE
add device discovery daemon

### LINKED ISSUE #1655: Discover devices independently of starting osd pods
For the [one osd per pod](https://github.com/rook/rook/blob/master/design/dedicated-osd-pod.md) design, the devices need to be discovered initially before any clusters are deployed. Then when a cluster crd is created, the operator can make decisions about what osd pods to deploy, one for each device. 

This will be a new daemon that would run on any node that is labeled to be used as a storage node.

When Rook starts consuming local storage PVs, this device discovery will become obsolete.

### PR DESCRIPTION
**Pull request by rootfs**
Description of your changes:

First step following up #1563. This PR implements the device discovery part. It creates the discovery daemon and scans the devices. Provisioning and rest will come in next PR.

@travisn @jbw976 

Which issue is resolved by this Pull Request:
Resolves #1655

Checklist:
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.

### CODE REVIEW

File: pkg/daemon/discover/discover.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		logger.Infof("failed to probe devices: %v", err)
+		return err
+	}
+	deviceJson, err := json.Marshal(devices)
+	if err != nil {
+		logger.Infof("failed to marshal: %v", err)
+		return err
+	}
+	deviceStr := string(deviceJson)
+	cmName := RawDeviceCMName + nodeName
```
Comment (travisn) [inline pkg/daemon/discover/discover.go]: might be nice to include a `-` between `devices` and `nodeName`

File: pkg/daemon/discover/discover.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+				Name:      cmName,
+				Namespace: namespace,
+				Labels: map[string]string{
+					k8sutil.AppAttr: AppName,
+					NodeAttr:        nodeName,
+				},
+			},
+			Data: make(map[string]string),
+		}
+		cm, err = context.Clientset.CoreV1().ConfigMaps(namespace).Create(cm)
```
Comment (travisn) [inline pkg/daemon/discover/discover.go]: is there a need to create with empty data before updating?

File: pkg/daemon/discover/discover.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		if device.Type == sys.PartType {
+			continue
+		}
+		ownPartition, devFS, err := sys.CheckIfDeviceAvailable(context.Executor, device.Name)
+		if err != nil {
+			logger.Infof("failed to check device %s: %v", device.Name, err)
+			continue
+		}
+		d := sys.RawDevice{
+			// FIXME: use persistent name
```
Comment (travisn) [inline pkg/daemon/discover/discover.go]: i think the best we can do here is the device name. As discussed, we can store the uuid of the partition when the osd actually consumes it.

File: pkg/daemon/discover/discover.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+		}
+		cm, err = context.Clientset.CoreV1().ConfigMaps(namespace).Create(cm)
+		if err != nil {
+			logger.Infof("failed to create configmap: %v", err)
+			return fmt.Errorf("failed to create raw device map %s: %+v", cmName, err)
+		}
+	}
+	if deviceStr != lastDevice {
+		data := make(map[string]string, 1)
+		data[RawDeviceCMData] = deviceStr
```
Comment (travisn) [inline pkg/daemon/discover/discover.go]: we will need to make sure other metadata about the devices does not get overwritten, such as the status of the devices related to orchestration. Will that status be stored in a different property of the configmap?

Comment (rootfs) [reply pkg/daemon/discover/discover.go]: yes, the orchestration status will be stored in a separate configmap.

File: pkg/clusterd/disk.go [hunk 5]
```diff
@@ -108,7 +108,8 @@ func DiscoverDevices(executor exec.Executor) ([]*LocalDisk, error) {
 
 		fs, err := sys.GetDeviceFilesystems(d, executor)
 		if err != nil {
-			return nil, err
+			logger.Warningf("skipping device %s with an unknown filesystem. %+v", d, err)
```
Comment (travisn) [inline pkg/clusterd/disk.go]: info level would be good, it doesn't seem like something the user needs to worry about

File: pkg/operator/discover/discover_test.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:  "mypodContainer",
+					Image: "rook/test",
+				},
+			},
+		},
+	}
+	clientset.CoreV1().Pods("rook-system").Create(&pod)
```
Comment (travisn) [inline pkg/operator/discover/discover_test.go]: what is tested by starting the mock operator pod? Seems like the pod is not needed for the validation in this test

File: pkg/util/sys/device.go [hunk 7]
```diff
@@ -143,12 +149,12 @@ func GetDevicePropertiesFromPath(devicePath string, executor exec.Executor) (map
 // get the file systems availab
 func GetDeviceFilesystems(device string, executor exec.Executor) (string, error) {
 	cmd := fmt.Sprintf("get filesystem type for %s", device)
-	output, err := executor.ExecuteCommandWithOutput(false, cmd, "df", "--output=source,fstype")
+	output, err := executor.ExecuteCommandWithOutput(false, cmd, "wipefs", "-p", "-n", fmt.Sprintf("/dev/%s", device))
```
Comment (travisn) [inline pkg/util/sys/device.go]: wipefs will detect the fs whether or not it is mounted?

Comment (rootfs) [reply pkg/util/sys/device.go]: yes, it reads off disk

File: pkg/clusterd/disk.go (hunk 5 above)
Comment (rootfs) [reply pkg/clusterd/disk.go]: this follows the other message severity.

File: pkg/operator/discover/discover_test.go (hunk 6 above)
Comment (rootfs) [reply pkg/operator/discover/discover_test.go]: done

**Comment by rootfs** [comment]
tested on my local host with the following configmap:

```
# kubectl describe cm raw-device-127.0.0.1 -n rook-system  
```
```yaml
apiVersion: v1
data:
  devices: '[{"devicePath":"/dev/sda","size":249510756352,"uuid":"","serial":"36b8ca3a0ed8348001a45349c07e33af9","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":false,"filesystem":"","empty":true,"extra":""},{"devicePath":"/dev/sdb","size":3000034656256,"uuid":"","serial":"36b8ca3a0ed8348001a4534a10823d858","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":false,"filesystem":"","empty":true,"extra":""},{"devicePath":"/dev/sdc","size":3000034656256,"uuid":"","serial":"36b8ca3a0ed8348001a4534a908aa65bb","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":false,"filesystem":"","empty":true,"extra":""},{"devicePath":"/dev/sdd","size":3000034656256,"uuid":"","serial":"36b8ca3a0ed8348001a4534ae08edd068","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":false,"filesystem":"","empty":true,"extra":""},{"devicePath":"/dev/sde","size":3000034656256,"uuid":"","serial":"36b8ca3a0ed8348001a4534b2092f30f1","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":false,"filesystem":"","empty":true,"extra":""},{"devicePath":"/dev/sdf","size":3000034656256,"uuid":"","serial":"36b8ca3a0ed8348001a4534b60970afde","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":false,"filesystem":"","empty":true,"extra":""},{"devicePath":"/dev/sdg","size":3000034656256,"uuid":"","serial":"36b8ca3a0ed8348001a4534bf09fa05fe","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":false,"filesystem":"","empty":true,"extra":""},{"devicePath":"/dev/sdh","size":3000034656256,"uuid":"","serial":"36b8ca3a0ed8348001a4534c40a3d1592","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":false,"filesystem":"","empty":true,"extra":""},{"devicePath":"/dev/sdi","size":3000034656256,"uuid":"","serial":"36b8ca3a0ed8348001a4534cd0ac5ce59","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":false,"filesystem":"","empty":true,"extra":""},{"devicePath":"/dev/sdj","size":3000034656256,"uuid":"","serial":"36b8ca3a0ed8348001a4534d10b0be139","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":false,"filesystem":"","empty":true,"extra":""},{"devicePath":"/dev/sdk","size":10737418240,"uuid":"f2d38cba-37da-411d-b7ba-9a6696c58174","serial":"36001405d27e5d898829468b90ce4ef8c","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":true,"filesystem":"ext2","empty":false,"extra":""},{"devicePath":"/dev/sdl","size":10737418240,"uuid":"","serial":"360014050d33c51c190a4e88a6664e234","type":"disk","rotational":true,"readOnly":false,"removable":false,"ownPartition":true,"filesystem":"","empty":true,"extra":""}]'
kind: ConfigMap
metadata:
  creationTimestamp: 2018-04-25T20:40:53Z
  labels:
    app: rook-discover
    rook.io/node: 127.0.0.1
  name: raw-device-127.0.0.1
  namespace: rook-system
  resourceVersion: "7159"
  selfLink: /api/v1/namespaces/rook-system/configmaps/raw-device-127.0.0.1
  uid: f2c1a511-48c8-11e8-ad5f-b8ca3a627e8c
```

File: pkg/operator/discover/discover.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+		}
+		logger.Infof("rook-discover daemonset already exists")
+	} else {
+		logger.Infof("rook-discover daemonset started")
+	}
+	return nil
+
+}
+
+func ListDevices(context *clusterd.Context, namespace, nodeName string) (map[string][]sys.RawDevice, error) {
```
Comment (ganeshmaharaj) [inline pkg/operator/discover/discover.go]: Pardon the dumbness of the question, but i don't see this being used anywhere in the code right now. is this for the provisioner to gather devices on a host and prepare them for osds?

Comment (rootfs) [reply pkg/operator/discover/discover.go]: yes, that's the 2nd step: rook operator will list devices using this function on each node and filter the devices. The filtered devices are then provisioned to become osds

File: pkg/operator/cluster/ceph/osd/osd.go [hunk 9]
```diff
@@ -168,9 +169,16 @@ func (c *Cluster) Start() error {
 			errorMessages = append(errorMessages, fmt.Sprintf("failed to set orchestration starting status for node %s: %+v", n.Name, err))
 			continue
 		}
-
+		devicesToUse := n.Devices
```
Comment (rootfs) [inline pkg/operator/cluster/ceph/osd/osd.go]: @ganeshmaharaj here you go

File: pkg/daemon/discover/discover.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+	NodeAttr        = "rook.io/node"
+	LocalDiskCMData = "devices"
+	LocalDiskCMName = "local-device-"
+)
+
+func Run(context *clusterd.Context) error {
+	if context == nil {
+		return fmt.Errorf("nil context")
+	}
+	nodeName := os.Getenv(k8sutil.NodeNameEnvVar)
```
Comment (travisn) [inline pkg/daemon/discover/discover.go]: the variables (nodeName and namespace) passed to the daemons can be loaded simply with the cobra command line options. Then you can also check that they are set before the daemon continues. Or if done this way, you may want to return an error if not set

File: pkg/daemon/discover/discover.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+		cm, err = context.Clientset.CoreV1().ConfigMaps(namespace).Create(cm)
+		if err != nil {
+			logger.Infof("failed to create configmap: %v", err)
+			return fmt.Errorf("failed to create local device map %s: %+v", cmName, err)
+		}
+		lastDevice = deviceStr
+	}
+	if deviceStr != lastDevice {
+		data := make(map[string]string, 1)
+		data[LocalDiskCMData] = deviceStr
```
Comment (travisn) [inline pkg/daemon/discover/discover.go]: the status of how rook is consuming the devices will be stored in a different key of the same configmap in the next PR? For now it's just raw properties of the devices?

File: pkg/daemon/discover/discover.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+			logger.Infof("failed to update configmap %s: %v", cmName, err)
+			return err
+		}
+	}
+
+	sigc := make(chan os.Signal, 1)
+	signal.Notify(sigc, syscall.SIGTERM)
+	for {
+		select {
+		case <-sigc:
```
Comment (travisn) [inline pkg/daemon/discover/discover.go]: should we periodically search for device changes as part of this loop?

File: pkg/operator/cluster/ceph/osd/osd.go [hunk 13]
```diff
@@ -168,9 +169,16 @@ func (c *Cluster) Start() error {
...
+		availDev, deviceErr := discover.GetAvailableDevices(c.context, n.Name, c.Namespace, n.Devices, n.Selection.DeviceFilter, n.Selection.GetUseAllDevices())
+		if deviceErr != nil {
+			logger.Warningf("failed to get devices for node %s cluster %s: %v", n.Name, c.Namespace, deviceErr)
+		} else {
+			devicesToUse = availDev
+			logger.Infof("avail devices for node %s: %+v", n.Name, availDev)
+		}
 		// create the replicaSet that will run the OSDs for this node
-		rs := c.makeReplicaSet(n.Name, n.Devices, n.Selection, resources, n.Config)
+		rs := c.makeReplicaSet(n.Name, devicesToUse, n.Selection, resources, n.Config)
```
Comment (travisn) [inline pkg/operator/cluster/ceph/osd/osd.go]: does this need to pass `availDev`?

File: pkg/operator/cluster/ceph/osd/osd.go (hunk 9 above)
Comment (travisn) [reply pkg/operator/cluster/ceph/osd/osd.go]: `n.Devices` is the set of desired devices, right?

File: pkg/operator/cluster/ceph/osd/osd.go [hunk 14]
```diff
@@ -168,9 +169,16 @@ func (c *Cluster) Start() error {
 			errorMessages = append(errorMessages, fmt.Sprintf("failed to set orchestration starting status for node %s: %+v", n.Name, err))
 			continue
 		}
-
+		devicesToUse := n.Devices
+		availDev, deviceErr := discover.GetAvailableDevices(c.context, n.Name, c.Namespace, n.Devices, n.Selection.DeviceFilter, n.Selection.GetUseAllDevices())
+		if deviceErr != nil {
+			logger.Warningf("failed to get devices for node %s cluster %s: %v", n.Name, c.Namespace, deviceErr)
```
Comment (travisn) [inline pkg/operator/cluster/ceph/osd/osd.go]: shouldn't we fail the install if we don't know what is available?

File: pkg/util/sys/device.go [hunk 15]
```diff
@@ -254,31 +301,46 @@ func UnmountDevice(devicePath string, executor exec.Executor) error {
...
-	lines := strings.Split(output, "\n")
-	for _, line := range lines {
-		if strings.HasPrefix(line, fmt.Sprintf("/dev/%s", device)) {
-			words := strings.Split(line, " ")
-			fs = append(fs, words[len(words)-1])
+func RookOwnsPartitions(partitions []*Partition) bool {
+
+	// if there are partitions, they must all have the rook osd label
+	for _, p := range partitions {
+		if !strings.HasPrefix(p.Label, "ROOK-OSD") {
```
Comment (travisn) [inline pkg/util/sys/device.go]: is `ROOK-OSD` defined in a const somewhere?

File: pkg/util/sys/device.go [hunk 16]
```diff
@@ -42,6 +41,44 @@ type Partition struct {
...
+	// Serial is the disk serial used by /dev/disk/by-id
+	Serial string `json:"serial"`
+	// Type is disk type
+	Type string `json:"type"`
+	// Rotational is the boolean whether the device is rotational: true for hdd, false for ssd and nvme
+	Rotational bool `json:"rotational"`
+	// ReadOnly is the boolean whether the device is readonly
+	Readonly bool `json:"readOnly"`
+	// OwnPartition is whether rook owns the partition
+	OwnPartition bool `json:"ownPartition"`
```
Comment (travisn) [inline pkg/util/sys/device.go]: when i tested the branch, `OwnPartition` was always false even on devices where the osd was configured with bluestore partitions.

File: pkg/util/sys/device.go [hunk 17]
```diff
@@ -42,6 +41,44 @@ type Partition struct {
...
+	// Filesystem is the filesystem currently on the device
+	Filesystem string `json:"filesystem"`
```
Comment (travisn) [inline pkg/util/sys/device.go]: the `filesystem` property was blank in the configmap even for my root drive

File: pkg/operator/cluster/ceph/osd/osd.go (hunk 9 above)
Comment (rootfs) [reply pkg/operator/cluster/ceph/osd/osd.go]: yes

File: pkg/daemon/discover/discover.go (hunk 11 above)
Comment (rootfs) [reply pkg/daemon/discover/discover.go]: yes, the device consumption will be made into a different configmap. Once the devices are used for one cluster, they will appear in the consumption configmap.  Operator lists both configmaps andd does an outer join to find the unused devices for new clusters.

File: pkg/operator/cluster/ceph/osd/osd.go (hunk 13 above)
Comment (rootfs) [reply pkg/operator/cluster/ceph/osd/osd.go]: this is done in the provisioning PR

File: pkg/operator/cluster/ceph/osd/osd.go (hunk 14 above)
Comment (rootfs) [reply pkg/operator/cluster/ceph/osd/osd.go]: there could be data dir path too. We fail only when neither device nor data dir are available.

File: pkg/util/sys/device.go (hunk 15 above)
Comment (rootfs) [reply pkg/util/sys/device.go]: good catch! the string seems to be everywhere, will do a cleanup in a separate PR.

File: pkg/util/sys/device.go (hunk 17 above)
Comment (rootfs) [reply pkg/util/sys/device.go]: if the root device has partitions then there is no filesystem

File: pkg/util/sys/device.go (hunk 16 above)
Comment (rootfs) [reply pkg/util/sys/device.go]: i see, `blkid` needs to run in privileged mode to read partition label.

Comment (rootfs) [reply pkg/util/sys/device.go]: i'll use udevadm to read parititon label

Comment (rootfs) [reply pkg/util/sys/device.go]: sample output
```
[root@rook-discover-lc57t /]# blkid /dev/sdc3 -s PARTLABEL -o value
[root@rook-discover-lc57t /]# udevadm info /dev/sdc3 |grep ROOK    
S: disk/by-partlabel/ROOK-OSD2-BLOCK
E: DEVLINKS=/dev/disk/by-id/scsi-3600140568c0bd28d4ee43769387c9f02-part3 /dev/disk/by-id/wwn-0x600140568c0bd28d4ee43769387c9f02-part3 /dev/disk/by-partlabel/ROOK-OSD2-BLOCK /dev/disk/by-partuuid/2089640e-bdeb-4fb4-aaec-88e165780b88 /dev/disk/by-path/ip-127.0.0.1:3260-iscsi-iqn.2016-06.world.srv:storage.target01-lun-2-part3
E: ID_PART_ENTRY_NAME=ROOK-OSD2-BLOCK
E: PARTNAME=ROOK-OSD2-BLOCK
```

File: pkg/daemon/discover/discover.go (hunk 12 above)
Comment (rootfs) [reply pkg/daemon/discover/discover.go]: done

File: pkg/util/sys/device.go [hunk 18]
```diff
@@ -63,6 +63,8 @@ type LocalDisk struct {
 	Rotational bool `json:"rotational"`
 	// ReadOnly is the boolean whether the device is readonly
 	Readonly bool `json:"readOnly"`
+	// Partitions is a partition slice
+	Partitions []Partition
```
Comment (travisn) [inline pkg/util/sys/device.go]: If we're going to return the partitions, is there a need for the OwnPartition property anymore?

File: pkg/daemon/discover/discover.go [hunk 19]
```diff
@@ -35,19 +36,41 @@ import (
...
 
 func Run(context *clusterd.Context) error {
 	if context == nil {
 		return fmt.Errorf("nil context")
 	}
-	nodeName := os.Getenv(k8sutil.NodeNameEnvVar)
-	namespace := os.Getenv(k8sutil.PodNamespaceEnvVar)
+	nodeName = os.Getenv(k8sutil.NodeNameEnvVar)
+	namespace = os.Getenv(k8sutil.PodNamespaceEnvVar)
+	cmName = LocalDiskCMName + nodeName
```
Comment (travisn) [inline pkg/daemon/discover/discover.go]: do we calculate the cmName in a few different places? maybe a helper function to make sure they don't get out of sync?

Comment (rootfs) [reply pkg/daemon/discover/discover.go]: this is the only place cmName is calculated.

File: pkg/util/sys/device.go (hunk 18 above)
Comment (rootfs) [reply pkg/util/sys/device.go]: good point, ownPartition is now removed

**Pull request merged**